### PR TITLE
Fuzz base64/v7

### PR DESCRIPTION
**Pull request by inashivb**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/6050

Previous PR: https://github.com/OISF/suricata/pull/10763

Changes since v6:
- rebased on top of latest master so the CI is green

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 19941

File: src/tests/fuzz/fuzz_decodebase64.c [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    uint8_t shift = 0;
+    uint32_t dest_size = 0;
+    if (size > BLK_SIZE)
+        shift = BIT_SHIFT_SIZE;
+    else
+        shift = BYTE_SIZE * (uint8_t)(size - 1);
+
+    for (size_t i = 0; i < size; i++) {
```
Comment (catenacyber) [inline src/tests/fuzz/fuzz_decodebase64.c:49]: `i < size && i < BLK_SIZE` in the for condition then ;-)

File: src/tests/fuzz/fuzz_decodebase64.c [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        uint32_t consumed_bytes = 0;
+        uint32_t decoded_bytes = 0;
+
+        DecodeBase64(dest, dest_size, src, len, &consumed_bytes, &decoded_bytes, mode);
+    }
+}
+
+#define BIT_SHIFT_SIZE 24
+#define BLK_SIZE       4
+#define BYTE_SIZE      8
```
Comment (catenacyber) [inline src/tests/fuzz/fuzz_decodebase64.c:28]: Need a better name, and I guess 2 is a better value

File: src/tests/fuzz/fuzz_decodebase64.c [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        // global init
+        InitGlobal();
+        run_mode = RUNMODE_UNITTEST;
+        initialized = 1;
+    }
+
+    uint8_t shift = 0;
+    uint32_t dest_size = 0;
+    if (size > BLK_SIZE)
+        shift = BIT_SHIFT_SIZE;
```
Comment (catenacyber) [inline src/tests/fuzz/fuzz_decodebase64.c:45]: I do not understand this shift logic...

File: src/tests/fuzz/fuzz_decodebase64.c [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        shift = BYTE_SIZE * (uint8_t)(size - 1);
+
+    for (size_t i = 0; i < size; i++) {
+        if (i == BLK_SIZE)
+            break;
+        dest_size |= (uint32_t)(data[i] << shift);
+        shift -= BYTE_SIZE;
+    }
+
+    Base64FuzzTest(data, size, dest_size);
```
Comment (catenacyber) [inline src/tests/fuzz/fuzz_decodebase64.c:56]: data needs to be increased and size decreased

like `Base64FuzzTest(data + BLK_SIZE, size - BLK_SIZE, dest_size);`

**Comment by catenacyber** [comment]
Maybe this requires CIFuzz fixing first cf https://github.com/google/oss-fuzz/pull/11797

File: src/tests/fuzz/fuzz_decodebase64.c [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+static int initialized = 0;
+
+static void Base64FuzzTest(const uint8_t *src, size_t len, size_t dest_size)
+{
+    for (uint8_t mode = BASE64_MODE_RELAX; mode <= BASE64_MODE_RFC4648; mode++) {
+        uint8_t *dest = malloc(dest_size);
+        uint32_t consumed_bytes = 0;
+        uint32_t decoded_bytes = 0;
+
+        DecodeBase64(dest, dest_size, src, len, &consumed_bytes, &decoded_bytes, mode);
```
Comment (catenacyber) [inline src/tests/fuzz/fuzz_decodebase64.c:22]: `dest` needs to be freed

**Review by catenacyber — CHANGES_REQUESTED** [review]
At least the leak by the fuzz target needs to be fixed

File: src/tests/fuzz/fuzz_decodebase64.c (hunk 3 above)
Comment (inashivb) [reply src/tests/fuzz/fuzz_decodebase64.c:45]: See the following:
1. Read the number of bytes we receive from the fuzzer.
2. If the bytes received from fuzzer >= 4, take first 4 bytes, shift them one by one by 24, 16, 8, 0 to make a u32 size.
3. If the bytes received from fuzzer < 4, take whatever bytes there are and try to make a random size out of it.

File: src/tests/fuzz/fuzz_decodebase64.c (hunk 1 above)
Comment (inashivb) [reply src/tests/fuzz/fuzz_decodebase64.c:49]: That would be cleaner indeed. Thanks! :)

File: src/tests/fuzz/fuzz_decodebase64.c (hunk 4 above)
Comment (inashivb) [reply src/tests/fuzz/fuzz_decodebase64.c:56]: Having trouble understanding this. Why should the data or size be tampered with? do we not receive random data and size from the fuzzer?

File: src/tests/fuzz/fuzz_decodebase64.c (hunk 3 above)
Comment (catenacyber) [reply src/tests/fuzz/fuzz_decodebase64.c:45]: Looks too complex : I would expect a fixed number of bytes for the length

File: src/tests/fuzz/fuzz_decodebase64.c (hunk 4 above)
Comment (catenacyber) [reply src/tests/fuzz/fuzz_decodebase64.c:56]: `data` is not tampered with : we do not modify the bytes inside the buffer.
But `data` is deserialized into 2 fields
- dest_size
- base64 data for the remaining part

These 2 fields should not overlap to be independent

Comment (inashivb) [reply src/tests/fuzz/fuzz_decodebase64.c:56]: ooh. Thank you!

File: src/tests/fuzz/fuzz_decodebase64.c (hunk 3 above)
Comment (inashivb) [reply src/tests/fuzz/fuzz_decodebase64.c:45]: ok. Keeping two bytes for length

**Pull request closed without merging**